server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesParams.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.mapper;
+
+import java.util.Arrays;
+import java.util.EnumSet;
+import java.util.function.Function;
+
+/**
+ * Utility functions for time series related mapper parameters
+ */
+public final class TimeSeriesParams {
+
+    public static final String TIME_SERIES_METRIC_PARAM = "time_series_metric";
+
+    private TimeSeriesParams() {}
+
+    public enum MetricType {
+        gauge,
+        counter,
+        histogram,
+        summary
```

Comment:
I'd make them `GUAGE` and `COUNTER` and `HISTOGRAM` just to follow along with normal java stuff. You'd have to convert to upper case on parse but I think that's fine.

Reply:
I started writing this change, but it turned out that code became a lot more complicated just to convert input to upper case and also convert enum values to lower case when logging error messages. Otherwise, user receives error messages like the following:

`"Unknown value [histogram] for field [time_series_metric] - accepted values are [GAUGE, COUNTER]"` 

So, sticking to enum constants same as the expected input values would result in a a lot simpler code.

We have followed the same approach [here](https://github.com/elastic/elasticsearch/blob/5baabff6670a8ed49297488ca8cac8ec12a2078d/x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java#L93-L101)

Reply:
Oh well. I don't like it there either! But I think it's cool - they can stay lowercase.

Reply:
It can also become relatively simple  if I replace the generic implementation (`restrictedEnumParam()`) from `FieldMapper` class and only add the custom implementation for the `MetricType` enum in the `TimeSeriesParams`.

Is a more specific to `MetricType` implementation with capitalized names  a better approach here?

Reply:
Sorry I don't understand - do you mean that you can either share the enum infrastructure or capitalize the names of the constants? If that's the case I think I prefer sharing the enum infrastructure. That way if we come up with a clever way to capitalize the names we can get it in a follow up and hit everything using the infrastructure.